src/MatrixPostMessageApi.js

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright 2017 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import Promise from "bluebird";
+
+function defer() {
```

Comment:
Deferring is considered an anti-pattern by bluebird/promises, although I can see why it's being used here. To avoid having this hidden away in the class I'd recommend one of two things:
1. Break it out to a utility somewhere (so it can be reused/maintained)
2. Avoid using defer. This can be achieved using a more minimal interface (described below)

Bear in mind, this is heavily based on what Dimension does: https://github.com/turt2live/matrix-dimension/blob/travis/ui-refresh/web/app/shared/services/scalar/scalar-client-api.service.ts

```
const requestMap = {}; // request_id: { resolve, reject }

// When a new request comes in:
new Promise((resolve, reject) => {
  requestMap[requestId] = { resolve, reject }; // requestId is whatever can be considered unique
});

// When you want to finalize a request
const promise = requestMap[requestId];
delete requestMap[requestId];
promise.resolve();
```

It's still a deferred, just without a 20ish line method to do it. The timeout can be done with a simple null check on `requestMap[requestId]`

Reply:
Thanks @turt2live - That is much cleaner 👍